**IMPORTING REQUIRED LIBRARIES**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
from sklearn.preprocessing import MinMaxScaler, LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, f1_score

In [3]:
! pip install scikeras

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [4]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.optimizers import Adam
from scikeras.wrappers import KerasClassifier

**LOADING DATASET**

In [6]:
df=pd.read_csv('/content/magic_gamma_telescope04_.csv')

In [38]:
df.head()

flength    fwidth   fsize   fconc  fconc1      fsym  fm3long  fm3trans  \
0   28.7967   16.0021  2.6449  0.3918  0.1982   27.7004  22.0110   -8.2027   
1   31.6036   11.7235  2.5185  0.5303  0.3773   26.2722  23.8238   -9.9574   
2  162.0520  136.0310  4.0612  0.0374  0.0187  116.7410 -64.8580  -45.2160   
3   23.8172    9.5728  2.3385  0.6147  0.3922   27.2107  -6.4633   -7.1513   
4   75.1362   30.9205  3.1611  0.3168  0.1832   -5.5277  28.5525   21.8393   

    falpha      dist class  
0  40.0920   81.8828     g  
1   6.3609  205.2610     g  
2  76.9600  256.7880     g  
3  10.4490  116.7370     g  
4   4.6480  356.4620     g

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19020 entries, 0 to 19019
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   flength   19020 non-null  float64
 1   fwidth    19020 non-null  float64
 2   fsize     19020 non-null  float64
 3   fconc     19020 non-null  float64
 4   fconc1    19020 non-null  float64
 5   fsym      19020 non-null  float64
 6   fm3long   19020 non-null  float64
 7   fm3trans  19020 non-null  float64
 8   falpha    19020 non-null  float64
 9   dist      19020 non-null  float64
 10  class     19020 non-null  object 
dtypes: float64(10), object(1)
memory usage: 1.6+ MB


In [40]:
df.isna().sum()

flength     0
fwidth      0
fsize       0
fconc       0
fconc1      0
fsym        0
fm3long     0
fm3trans    0
falpha      0
dist        0
class       0
dtype: int64

In [41]:
df.describe()

flength        fwidth         fsize         fconc        fconc1  \
count  19020.000000  19020.000000  19020.000000  19020.000000  19020.000000   
mean      53.250154     22.180966      2.825017      0.380327      0.214657   
std       42.364855     18.346056      0.472599      0.182813      0.110511   
min        4.283500      0.000000      1.941300      0.013100      0.000300   
25%       24.336000     11.863800      2.477100      0.235800      0.128475   
50%       37.147700     17.139900      2.739600      0.354150      0.196500   
75%       70.122175     24.739475      3.101600      0.503700      0.285225   
max      334.177000    256.382000      5.323300      0.893000      0.675200   

               fsym       fm3long      fm3trans        falpha          dist  
count  19020.000000  19020.000000  19020.000000  19020.000000  19020.000000  
mean      -4.331745     10.545545      0.249726     27.645707    193.818026  
std       59.206062     51.000118     20.827439     26.103621     74.731787  
min     -457.916100   -331.780000   -205.894700      0.000000      1.282600  
25%      -20.586550    -12.842775    -10.849375      5.547925    142.492250  
50%        4.013050     15.314100      0.666200     17.679500    191.851450  
75%       24.063700     35.837800     10.946425     45.883550    240.563825  
max      575.240700    238.321000    179.851000     90.000000    495.561000

**SPLITTING TARGET AND FEATURES**

In [7]:
X = df.drop('class', axis=1).values
y = df['class'].values

**TRAIN TEST SPLIT**

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

**MIN MAX SCALING**

In [9]:
scale = MinMaxScaler()
X_train =scale.fit_transform(X_train)
X_test = scale.fit_transform(X_test)

**LABEL ENCODING**

In [10]:
le = LabelEncoder()
y_en = le.fit_transform(y)
y_train_en = le.fit_transform(y_train)
y_test_en = le.fit_transform(y_test)

**MODEL**

In [11]:
model = Sequential()

model.add(Dense(16, input_shape=(10,), activation='relu'))
model.add(Dense(16, activation='relu'))

model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [13]:
model.fit(X_train, y_train_en, batch_size=12, epochs=5, 
          validation_data=(X_test, y_test_en))

Epoch 1/5
1268/1268 [==============================] - 4s 2ms/step - loss: 0.4803 - accuracy: 0.7717 - val_loss: 0.4431 - val_accuracy: 0.7989
Epoch 2/5
1268/1268 [==============================] - 4s 3ms/step - loss: 0.4183 - accuracy: 0.8094 - val_loss: 0.4168 - val_accuracy: 0.8191
Epoch 3/5
1268/1268 [==============================] - 3s 2ms/step - loss: 0.4032 - accuracy: 0.8172 - val_loss: 0.3991 - val_accuracy: 0.8278
Epoch 4/5
1268/1268 [==============================] - 3s 2ms/step - loss: 0.3930 - accuracy: 0.8253 - val_loss: 0.4061 - val_accuracy: 0.8333
Epoch 5/5
1268/1268 [==============================] - 3s 2ms/step - loss: 0.3857 - accuracy: 0.8339 - val_loss: 0.3951 - val_accuracy: 0.8370


In [15]:
predictions = (model.predict(X_test) > 0.5).astype('int32')

119/119 [==============================] - 0s 1ms/step


In [37]:
accuracy_score(y_test_en, predictions)

0.7970136698212408


In [18]:
f1_score(y_test_en, predictions)

0.7622699386503067

model gave 79.70% accuracy and 76.22% f-1 score.

**HYPER PARAMETER TUNING**

In [31]:
def create_model(neurons, optimizer):

	model = Sequential()
	model.add(Dense(neurons, input_shape=(10,), activation='relu'))
	model.add(Dense(neurons, activation='relu'))
	model.add(Dense(1, activation='sigmoid'))
 
	model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
	return model

In [32]:
optimizer = ['SGD', 'RMSprop', 'Adam']
neurons = [ 16, 32, 64]
Epochs=[10,25,35]
param_grid = dict(model__optimizer=optimizer, 
                  model__neurons=neurons,epochs=Epochs)

In [33]:
model = KerasClassifier(model=create_model, epochs=Epochs, batch_size=10, verbose=0)

In [34]:
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-1, cv=2)


In [35]:
grid_result = grid.fit(X, y_en)

/usr/local/lib/python3.7/dist-packages/joblib/externals/loky/process_executor.py:703: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory leak.", UserWarning


In [36]:

print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Best: 0.833228 using {'epochs': 35, 'model__neurons': 64, 'model__optimizer': 'Adam'}


We are able to increase accuracy up to 89.32% through hyper parameter tuning